File: tests/NTIX.Tests/PackageManagerDetectorValidationTests.cs
using System.Collections.Generic;
using FluentAssertions;
using NTIX.Core.Models;
using NTIX.Core.PackageManager;

namespace NTIX.Tests;

public class PackageManagerDetectorValidationTests
{
    [Fact]
    public void ValidateManagers_ChocolateyEnabledNotInstalled_ReturnsError()
    {
        if (PackageManagerDetector.IsChocolateyInstalled())
            return; // Skip if choco is installed

        var options = new NTIXOptions(
            new WingetOptions(),
            new ChocoOptions(Enable: true),
            new ScoopOptions());
        var config = new NTIXConfig(options);
        
        var (valid, error, warnings) = PackageManagerDetector.ValidateManagers(options, config);
        
        valid.Should().BeFalse();
        error.Should().Contain("Chocolatey is enabled but not installed");
        error.Should().Contain("chocolatey.org/install");
    }

    [Fact]
    public void ValidateManagers_ScoopEnabledNotInstalled_ReturnsError()
    {
        if (PackageManagerDetector.IsScoopInstalled())
            return; // Skip if scoop is installed

        var options = new NTIXOptions(
            new WingetOptions(),
            new ChocoOptions(),
            new ScoopOptions(Enable: true));
        var config = new NTIXConfig(options);
        
        var (valid, error, warnings) = PackageManagerDetector.ValidateManagers(options, config);
        
        valid.Should().BeFalse();
        error.Should().Contain("Scoop is enabled but not installed");
        error.Should().Contain("scoop.sh");
    }

    [Fact]
    public void ValidateManagers_ChocolateyPackagesNotEnabled_ReturnsWarning()
    {
        var options = new NTIXOptions(
            new WingetOptions(),
            new ChocoOptions(Enable: false),
            new ScoopOptions());
        var config = new NTIXConfig(
            options,
            ChocoPackages: new List<PackageEntry> { new("git", "1.0") });
        
        var (valid, error, warnings) = PackageManagerDetector.ValidateManagers(options, config);
        
        valid.Should().BeTrue();
        error.Should().BeNull();
        warnings.Should().Contain("[warn] Chocolatey packages declared but chocolatey not enabled in options");
    }

    [Fact]
    public void ValidateManagers_ScoopPackagesNotEnabled_ReturnsWarning()
    {
        var options = new NTIXOptions(
            new WingetOptions(),
            new ChocoOptions(),
            new ScoopOptions(Enable: false));
        var config = new NTIXConfig(
            options,
            ScoopPackages: new List<PackageEntry> { new("fd", "1.0") });
        
        var (valid, error, warnings) = PackageManagerDetector.ValidateManagers(options, config);
        
        valid.Should().BeTrue();
        error.Should().BeNull();
        warnings.Should().Contain("[warn] Scoop packages declared but scoop not enabled in options");
    }

    [Fact]
    public void ValidateManagers_AllDisabledNoPackages_ReturnsSuccess()
    {
        var options = new NTIXOptions();
        var config = new NTIXConfig(options);
        
        var (valid, error, warnings) = PackageManagerDetector.ValidateManagers(options, config);
        
        valid.Should().BeTrue();
        error.Should().BeNull();
        warnings.Should().BeEmpty();
    }

    [Fact]
    public void ValidateManagers_NullOptions_HandlesGracefully()
    {
        var options = new NTIXOptions();
        var config = new NTIXConfig(options);
        
        var (valid, error, warnings) = PackageManagerDetector.ValidateManagers(options, config);
        
        valid.Should().BeTrue();
        error.Should().BeNull();
    }
}

File: tests/package_manager_detector_validation_tests.rs
use ntix_rs::models::ntix_config::NTIXConfig;
use ntix_rs::models::options::{ChocoOptions, NTIXOptions, ScoopOptions, WingetOptions};
use ntix_rs::models::package_entry::PackageEntry;
use ntix_rs::package_manager::package_manager_detector;

mod common;
use common::MockManagerPresence;

fn opts(choco_enable: bool, scoop_enable: bool) -> NTIXOptions {
    NTIXOptions {
        winget: WingetOptions::default(),
        chocolatey: ChocoOptions {
            enable: choco_enable,
            ..Default::default()
        },
        scoop: ScoopOptions {
            enable: scoop_enable,
            ..Default::default()
        },
    }
}

#[test]
fn validate_managers_chocolatey_enabled_not_installed_warns_and_continues() {
    let presence = MockManagerPresence::with_choco(false);

    let options = opts(true, false);
    let config = NTIXConfig {
        options: options.clone(),
        ..Default::default()
    };

    let result = package_manager_detector::validate_managers(&options, &config, Some(&presence));
    assert!(!result.choco_installed);
    assert!(result
        .warnings
        .iter()
        .any(|w| w.contains("Chocolatey is enabled but not installed")
            && w.contains("chocolatey.org/install")));
}

#[test]
fn validate_managers_chocolatey_enabled_installed_returns_valid() {
    let presence = MockManagerPresence::new();

    let options = opts(true, false);
    let config = NTIXConfig {
        options: options.clone(),
        ..Default::default()
    };

    let result = package_manager_detector::validate_managers(&options, &config, Some(&presence));
    assert!(result.choco_installed);
    assert!(result.warnings.is_empty());
}

#[test]
fn validate_managers_scoop_enabled_not_installed_warns_and_continues() {
    let mut presence = MockManagerPresence::new();
    presence.scoop_installed = false;

    let options = opts(false, true);
    let config = NTIXConfig {
        options: options.clone(),
        ..Default::default()
    };

    let result = package_manager_detector::validate_managers(&options, &config, Some(&presence));
    assert!(!result.scoop_installed);
    assert!(result
        .warnings
        .iter()
        .any(|w| w.contains("Scoop is enabled but not installed") && w.contains("scoop.sh")));
}

#[test]
fn validate_managers_scoop_enabled_installed_returns_valid() {
    let presence = MockManagerPresence::new();

    let options = opts(false, true);
    let config = NTIXConfig {
        options: options.clone(),
        ..Default::default()
    };

    let result = package_manager_detector::validate_managers(&options, &config, Some(&presence));
    assert!(result.scoop_installed);
    assert!(result.warnings.is_empty());
}

#[test]
fn validate_managers_chocolatey_packages_not_enabled_returns_warning() {
    let options = opts(false, false);
    let config = NTIXConfig {
        options: options.clone(),
        choco_packages: vec![PackageEntry {
            id: "git".to_string(),
            version: Some("1.0".to_string()),
        }],
        ..Default::default()
    };

    let result = package_manager_detector::validate_managers(&options, &config, None);
    assert!(result.warnings.iter().any(|w| w.contains(
        "[warn] Chocolatey packages declared but chocolatey not enabled in options"
    )));
}

#[test]
fn validate_managers_scoop_packages_not_enabled_returns_warning() {
    let options = opts(false, false);
    let config = NTIXConfig {
        options: options.clone(),
        scoop_packages: vec![PackageEntry {
            id: "fd".to_string(),
            version: Some("1.0".to_string()),
        }],
        ..Default::default()
    };

    let result = package_manager_detector::validate_managers(&options, &config, None);
    assert!(result.warnings.iter().any(|w| w.contains(
        "[warn] Scoop packages declared but scoop not enabled in options"
    )));
}

#[test]
fn validate_managers_all_disabled_no_packages_returns_success() {
    let options = NTIXOptions::default();
    let config = NTIXConfig {
        options: options.clone(),
        ..Default::default()
    };

    let result = package_manager_detector::validate_managers(&options, &config, None);
    assert!(result.warnings.is_empty());
}

#[test]
fn validate_managers_null_options_handles_gracefully() {
    let options = NTIXOptions::default();
    let config = NTIXConfig {
        options: options.clone(),
        ..Default::default()
    };

    package_manager_detector::validate_managers(&options, &config, None);
}
